manta-pay/src/crypto/hash/poseidon/parameter_generation/mds.rs

```diff
@@ -0,0 +1,412 @@
+
+//! MDS Data Generation
+
+use super::matrix::{MatrixOperations, SquareMatrix};
+use crate::crypto::hash::poseidon::{parameter_generation::matrix::Matrix, Field, FieldGeneration};
+use alloc::{vec, vec::Vec};
+use core::fmt::Debug;
+use manta_util::vec::VecExt;
+
+/// MDS Matrix for both naive poseidon hash and optimized poseidon hash
+/// For detailed descriptions, please refer to <https://hackmd.io/8MdoHwoKTPmQfZyIKEYWXQ>
+/// Note: Naive and optimized poseidon hash does not change #constraints in Groth16.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct MdsMatrices<F>
+where
+    F: Field,
+{
+    /// MDS Matrix for naive poseidon hash
+    pub m: SquareMatrix<F>,
+    /// inversion of mds matrix. Used in optimzed poseidon hash.
+    pub m_inv: SquareMatrix<F>,
+    /// m_hat matrix. Used in optimized poseidon hash
+    pub m_hat: SquareMatrix<F>,
+    /// Inversion of m_hat matrix. Used in optimized poseidon hash.
+    pub m_hat_inv: SquareMatrix<F>,
+    /// m prime matrix. Used in optimized poseidon hash.
+    pub m_prime: SquareMatrix<F>,
+    /// m double prime matrix. Used in optimized poseidon hash.
+    pub m_double_prime: SquareMatrix<F>,
+}
+
+impl<F> MdsMatrices<F>
+where
+    F: Field + Clone,
+{
+    fn make_v_w(m: &Matrix<F>) -> (Vec<F>, Vec<F>) {
+        let v = m[0][1..].to_vec();
+        let w = m
+            .iter_rows()
+            .skip(1)
+            .map(|column| column[0].clone())
+            .collect();
+        (v, w)
+    }
+}
+
+impl<F> MdsMatrices<F>
+where
+    F: Field + Clone,
+{
+    fn make_prime(m: &SquareMatrix<F>) -> SquareMatrix<F> {
+        SquareMatrix::new(
+            Matrix::new(
+                m.iter_rows()
+                    .enumerate()
+                    .map(|(i, row)| match i {
+                        0 => {
+                            let mut new_row = Vec::allocate_with(row.len(), F::zero);
+                            new_row[0] = F::one();
+                            new_row
+                        }
+                        _ => {
+                            let mut new_row = Vec::allocate_with(row.len(), F::zero);
+                            new_row[1..].clone_from_slice(&row[1..]);
+                            new_row
+                        }
+                    })
+                    .collect(),
+            )
+            .unwrap(),
+        )
+        .unwrap()
+    }
+}
+
+impl<F> MdsMatrices<F>
+where
+    F: Clone + Field + FieldGeneration + PartialEq,
+{
+    /// Derives MDS matrix of size `dim*dim` and relevant things
+    pub fn new(dim: usize) -> Option<Self> {
+        Self::generate_mds(dim).map(Self::derive_mds_matrices)
+    }
+
+    /// Generates the mds matrix `m` for naive poseidon hash.
+    /// mds matrix is constructed to be symmetry so that row-major or col-major
+    /// representation gives the same output
+    pub fn generate_mds(t: usize) -> Option<SquareMatrix<F>>
+    where
+        F: FieldGeneration,

```

Comment:
minimize requirements

Reply:
`FieldGeneration` is required for `F::from_u64`.